### PR TITLE
SourceControl: centralise git cloning

### PR DESCRIPTION
**Pull request by compnerd**
Centralise the git clone command invocation to ensure that we uniformly handle the arguments for the clone.  This sets the stage for further tweaks needed to fix cloning some repositories on Windows.

### CODE REVIEW

**Review by neonichu — APPROVED** [review]

**Comment by compnerd** [comment]
@swift-ci please smoke test

File: Sources/SourceControl/GitRepository.swift [hunk 1]
```diff
@@ -112,6 +130,17 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
+    @discardableResult
+    private func callGit(_ args: String...,
+                         environment: EnvironmentVariables = Git.environment,
+                         repository: RepositorySpecifier,
+                         failureMessage: String = "",
+                         progress: FetchProgress.Handler? = nil) throws -> String {
+        return try callGit(args.map { $0 }, environment: environment,
+                           repository: repository,
+                           failureMessage: failureMessage, progress: progress)
+    }
```
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:142]: formatting

File: Sources/SourceControl/GitRepository.swift [hunk 2]
```diff
@@ -122,18 +151,8 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
-        //       particularly important for Windows where the process can prevent future operations.
-        try self.callGit("clone",
-                         "-c", "core.symlinks=true",
-                         "-c", "core.fsmonitor=false",
-                         "--mirror", repository.location.gitURL, path.pathString,
-                         repository: repository,
-                         failureMessage: "Failed to clone repository \(repository.location)",
-                         progress: progressHandler)
+        try self.clone(repository, repository.location.gitURL, path.pathString,
+                       ["--mirror"], progress: progressHandler)
```
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:155]: formatting

File: Sources/SourceControl/GitRepository.swift [hunk 3]
```diff
@@ -180,11 +199,10 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
             // present in the bare repository.
-            //
-            // NOTE: Explicitly set `core.symlinks=true` on `git clone` to ensure that symbolic links are correctly resolved.
-            try self.callGit("clone", "-c", "core.symlinks=true", "--no-checkout", sourcePath.pathString, destinationPath.pathString,
-                             repository: repository,
-                             failureMessage: "Failed to clone repository \(repository.location)")
+            try self.clone(repository,
+                           sourcePath.pathString, destinationPath.pathString,
+                           ["--no-checkout"])
+
```
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:205]: formatting

File: Sources/SourceControl/GitRepository.swift [hunk 4]
```diff
@@ -202,11 +220,9 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
...
             // only ever expect to get back a revision that remains present in the
             // object storage.
-            //
-            // NOTE: Explicitly set `core.symlinks=true` on `git clone` to ensure that symbolic links are correctly resolved.
-            try self.callGit("clone", "-c", "core.symlinks=true", "--shared", "--no-checkout", sourcePath.pathString, destinationPath.pathString,
-                             repository: repository,
-                             failureMessage: "Failed to clone repository \(repository.location)")
+            try self.clone(repository,
+                           sourcePath.pathString, destinationPath.pathString,
+                           ["--shared", "--no-checkout"])
```
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:225]: formatting

**Comment by compnerd** [comment]
I'll rebuild SwiftFormat and run it through.  I really wish that we could run that in CI.

File: Sources/SourceControl/GitRepository.swift [hunk 5]
```diff
@@ -82,28 +82,46 @@ public struct GitRepositoryProvider: RepositoryProvider, Cancellable {
         self.git = GitShellHelper(cancellator: cancellator)
     }
 
+    private func clone(_ repository: RepositorySpecifier,
```
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:85]: can we move this function after the "fetch" function, so the callGit are the first functions

Comment (compnerd) [reply Sources/SourceControl/GitRepository.swift:85]: Sure

File: Sources/SourceControl/GitRepository.swift (hunk 1 above)
Comment (tomerd) [reply Sources/SourceControl/GitRepository.swift:142]: we have been asking folks to run `swiftformat` on changed code, see the "soundness" script

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]
thank you

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Pull request merged**